[fluent-bit] feat: Add ability to specify more rules inside NetworkPolicy

mkilchhofer opened the pull request:
Implemented 3 features around the existing NetworkPolicy:
- Ability to include provided extraPorts inside NetworkPolicy (`networkPolicy.ingress.includeExtraPorts`)
- Ability to provide custom egress rules inside NetworkPolicy (`networkPolicy.egress.customRules`)
- Ability to provide custom ingress rules inside NetworkPolicy (`networkPolicy.ingress.customRules`)

/cc: @AlexisMtr as you implemented basic NetworkPolicy support in https://github.com/fluent/helm-charts/pull/65

stevehipwell commented:
@mkilchhofer I'm AFK and will look properly next week, but at first glance it looks like this wouldn't work correctly if only egress rules were set?

stevehipwell commented:
@mkilchhofer it looks like you need to update the logic to support only egress rules as the previous logic relied on the fact that if this was enabled there would be ingress rules only.

mkilchhofer commented:
Rebased again.
@stevehipwell is there anything more to adapt?

stevehipwell commented:
@mkilchhofer you still have a major logic flaw. The current implementation is binary, off or ingress only, so the implementation is simple. You want to make this more complex so you need to add additional filters to make sure that it works correctly; there is no longer an implicit link between this being on and there being ingress rules.

Also the configuration/template pattern should be consistent for both ingress and egress.

mkilchhofer commented:
Hmmm I don't understand why it is not an on/off now with my changes.

Before my change:
1. on / off by setting `networkPolicy.enabled`
   -> Still supported and still `bool`
2. limit ingress traffic to a specific rule via `networkPolicy.ingress.from`
  -> Still supported

Now I simplified the feature for option 2 (with 100% backward compat) and added two more customizations:
a) extend ingress filtering with custom rules
b) add ability to add custom egress rules

stevehipwell commented on charts/fluent-bit/templates/networkpolicy.yaml:
```diff
@@ -7,16 +7,30 @@ metadata:
     {{- include "fluent-bit.labels" . | nindent 4 }}
 spec:
   policyTypes:
+    {{- if .Values.networkPolicy.egress.customRules }}
```
You need to add a conditional for ingress now as it is no longer implicitly wanted if `networkPolicy.enabled` is set `true`.

You should probably also extend the primary gate to something along the lines of the below statement.

```helm
{{- if and .Values.networkPolicy.enabled .Values.networkPolicy.ingress.customRules .Values.networkPolicy.egress.customRules }}
```

stevehipwell commented on charts/fluent-bit/values.yaml:
```diff
@@ -279,8 +279,30 @@ existingConfigMap: ""
-#   ingress:
-#     from: []
+  egress:
+    customRules: []
```
I think this should be `rules`.

stevehipwell commented on charts/fluent-bit/values.yaml:
```diff
@@ -279,8 +279,30 @@ existingConfigMap: ""
+    #     cidr: 172.17.0.0/16
+    #     except:
+    #       - 172.17.1.0/24
+    includeExtraPorts: false
```
I don't think this is needed, ports shouldn't be blindly added and `rules` can be manually configured.

stevehipwell reviewed (changes requested):
I've added comments on the code regarding my previous feedback.
